### PR TITLE
feat Support seam components

### PR DESCRIPTION
**Pull request by azat-co**

### CODE REVIEW

File: src/types/models.ts [hunk 1]
```diff
@@ -285,3 +285,11 @@ type Flatten<EventType extends SeamEvent["event_type"]> =
     : never
 
 export type Event = Flatten<SeamEvent["event_type"]>
+
+export interface ClientSessionI {
```
Comment (codetheweb) [inline src/types/models.ts:289]: 🐧 we don't generally add `I`/`T` suffixes in this repo

File: src/seam-connect/client.ts [hunk 2]
```diff
@@ -73,8 +80,8 @@ export class Seam extends Routes {
       baseURL: endpoint,
       headers: {
         ...axiosOptions?.headers,
-        Authorization: `Bearer ${apiKey}`,
-        ["User-Agent"]: `Javascript SDK v${version} (https://github.com/seamapi/javascript)`,
+        Authorization: `Bearer ${apiKey || clientAccessToken}`,
+        // ["User-Agent"]: `Javascript SDK v${version} (https://github.com/seamapi/javascript)`,
```
Comment (codetheweb) [inline src/seam-connect/client.ts:84]:
```suggestion
        ["User-Agent"]: `Javascript SDK v${version} (https://github.com/seamapi/javascript)`,
```

File: src/seam-connect/client.ts [hunk 3]
```diff
@@ -111,4 +118,51 @@ export class Seam extends Routes {
...
+    } catch (e: any) {
+      return {
+        ok: false,
+        error: {
+          type: "unknown",
+          message: e.toString() ?? "Something went wrong",
+        },
+      }
+    }
+  }
```
Comment (codetheweb) [inline src/seam-connect/client.ts:159]: I think we generally throw instead of directly returning the error

File: src/types/models.ts (hunk 1 above)
Comment (razor-x) [reply src/types/models.ts:289]: If we are going to introduce this, I think we should rename all interfaces to be consistent then.

File: src/seam-connect/client.ts (hunk 2 above)
Comment (razor-x) [reply src/seam-connect/client.ts:84]: What are the errors? I'm assuming the browser doesn't like when JavaScript tries to override the user agent header. If so we probably need to selectively disable this in the browser environment.

Comment (codetheweb) [reply src/seam-connect/client.ts:84]: yeah we use the user agent in general to track SDK usage

maybe we could do a `typeof window === "undefined"` check?

Comment (codetheweb) [reply src/seam-connect/client.ts:84]: maybe we could 
1. keep `user-agent` for now until tooling is updated
2. disable `user-agent` when in a browser
3. always send a second header like `x-seam-sdk-version` with the same info

File: src/seam-connect/client.ts [hunk 4]
```diff
@@ -111,4 +118,51 @@ export class Seam extends Routes {
...
+}
+
+type catParams = {
```
Comment (seveibar) [inline src/seam-connect/client.ts:162]: PascalCase types, need linter here

```suggestion
type CATParams = {
```

File: src/seam-connect/client.ts [hunk 5]
```diff
@@ -111,4 +118,51 @@ export class Seam extends Routes {
...
+      // backend mode
+      params["api_key"] = ops.apiKey
+    } else if (ops.pubKey?.startsWith("seam_pk")) {
+      // frontend mode
+      params["pub_key"] = ops.pubKey
+    }
+    params["user_identifier_key"] = ops.userIdentifierKey
+    try {
+      const response = await axios.post(
+        ops.endpoint + "internal/client_access_tokens/create",
```
Comment (seveibar) [inline src/seam-connect/client.ts:137]:
```suggestion
        ops.endpoint + "internal/client_access_tokens/get_or_create_empty",
```

Comment (seveibar) [reply src/seam-connect/client.ts:137]: it's more clear that it will reuse an existing client session this way

File: src/types/models.ts (hunk 1 above)
Comment (seveibar) [reply src/types/models.ts:289]: * we should match existing repo convention to avoid bikeshedding (same rules as open source), so shouldn't be there
* should be prefix rather than suffix to match other Seam repos that use this convention
* not sure there's a meaningful enough different between the interface and a type, the convention makes the most sense where classes are used. Since we almost never use classes (this SDK is a bit of an exception), it's usually unnecessary. Would be strongly opposed to adding a T prefix to types and to a lesser extent the I prefix, since it's typically already distinguished by PascalCase and there are no classes so that's enough
* If this comes up more often, linter convention is better

**Comment by seveibar** [comment]
I think the overall approach is good!

File: package.json [hunk 6]
```diff
@@ -7,7 +7,7 @@
     "iot",
     "device"
   ],
-  "version": "5.12.0",
+  "version": "5.12.1",
```
Comment (razor-x) [inline package.json:10]: How did this version change in this PR branch? This doesn't seems right.

File: src/seam-connect/client.ts [hunk 7]
```diff
@@ -111,4 +126,51 @@ export class Seam extends Routes {
       throw error
     }
   }
+
+  static async getClientSessionToken(
+    ops: CSTParams
```
Comment (razor-x) [inline src/seam-connect/client.ts:131]: Can we make this `CSTParams & SeamClientOptions`? This way users can just spread the same options whenever they call `new Seam` or `Seam.anyStaticMethod`?

File: src/seam-connect/client.ts [hunk 8]
```diff
@@ -1,13 +1,24 @@
...
+  SuccessfulAPIResponse,
+} from "../types/globals"
 import { version } from "../../package.json"
+import { ClientSessionResponseInterface } from "../types"
 
 export interface SeamClientOptions {
   /* Seam API Key */
   apiKey?: string
+  /* Seam Client Access Token */
+  clientAccessToken?: string
```
Comment (razor-x) [inline src/seam-connect/client.ts:21]: I noticed in the new seam provider, we just pass all the keys as `apiKey` even if it's a `clientAccessToken`.

Do we still want a new option? Or just have `apiKey`, let the user pass whatever token they have, and then use the prefix if we need to do special logic?

**Review by razor-x — COMMENTED** [review]
Some questions on the option passing

File: src/seam-connect/client.ts [hunk 9]
```diff
@@ -111,4 +126,51 @@ export class Seam extends Routes {
...
+        error.response.status,
+        error.response.headers["seam-request-id"],
+        (error.response.data as ErroredAPIResponse).error
+      )
+    }
+  }
+}
+
+type CSTParams = {
+  publishedKey?: string
```
Comment (razor-x) [inline src/seam-connect/client.ts:171]:
```suggestion
  publishableKey?: string
```

File: src/seam-connect/client.ts (hunk 7 above)
Comment (razor-x) [reply src/seam-connect/client.ts:131]: Actually, we might just need

```suggestion
    ops: SeamClientOptions & { userIdentifierKey?: string | undefined | null }
```

**Comment by razor-x** [comment]
@azat-co To make it easier to grok so we can discuss, here are my proposed interface changes illustrated visually.  

image

File: package.json (hunk 6 above)
Comment (razor-x) [reply package.json:10]: We should revert this. It will confuse semantic release. Versions are automatically released off of main when PRs are merged

File: src/types/models.ts [hunk 10]
```diff
@@ -285,3 +285,11 @@ type Flatten<EventType extends SeamEvent["event_type"]> =
     : never
 
 export type Event = Flatten<SeamEvent["event_type"]>
+
+export interface ClientSession {
+  client_access_token_id: string
```
Comment (seveibar) [inline src/types/models.ts:290]: /client access token/client session token/g

File: src/types/route-responses.ts [hunk 11]
```diff
@@ -122,3 +123,7 @@ export interface ActionAttemptGetResponse<T extends ActionType = ActionType> {
 export interface EventsListResponse {
   events: Event[]
 }
+
+export interface ClientSessionResponseInterface {
```
Comment (seveibar) [inline src/types/route-responses.ts:127]: Is there a reason for the Interface suffix? Not our usual convention

**Comment by seveibar** [comment]
@razor-x disagree with making `apiKey` represent resources (sort of a polymorphic opaque type) if that's the proposal, simplifies implementation slightly at the risk of confusion

File: src/seam-connect/client.ts [hunk 12]
```diff
@@ -111,4 +126,51 @@ export class Seam extends Routes {
...
+      )
+      if (response.data.error) {
+        return {
+          ok: false,
+          error: {
+            type: response.data.error.type,
+            message: response.data.error.message,
+          },
+        }
+      }
```
Comment (seveibar) [inline src/seam-connect/client.ts:158]:
```suggestion
```

File: src/seam-connect/client.ts (hunk 8 above)
Comment (seveibar) [reply src/seam-connect/client.ts:21]: i think we should be generous in acceptance (since it's detectable) but we shouldn't conflate the options

**Comment by razor-x** [comment]
> i think we should be generous in acceptance (since it's detectable) but we shouldn't conflate the options

@seveibar 
 
Yes so I think we should either avoid a new option and only have apiKey, or add a new option but not conflate them. So in that case if they try to provide both maybe that is an error, or if they pass in something with a mismatched prefix we should warn them.

**Comment by seveibar** [comment]
> Yes so I think we should either avoid a new option and only have apiKey, or add a new option but not conflate them. So in that case if they try to provide both maybe that is an error, or if they pass in something with a mismatched prefix we should warn them.

@razor-x yep agree, a warning/error that they need to use a different parameter is good. Could create an issue from this comment?

**Comment by razor-x** [comment]
@seveibar @azat-co Ok how about this for the public interface changes?

1. Add `{ clientSessionToken?: string }` to `SeamClientOptions`.
2. Throw an error if both `apiKey` and `clientSessionToken` are provided.
3. Emit a warning if the key prefix for `apiKey` or `clientSessionToken` do not match their expected prefix.
4. Update signature of `getClientSessionToken` to 

```ts
static async getClientSessionToken(
    ops: Omit<SeamClientOptions, "apiKey" | "clientAccessToken"> &  { 
           publishableKey: string
           clientuserIdentifierKey?: string | undefined | null
         }
): Promise<APIResponse<ClientSessionResponse>>
```

File: src/seam-connect/client.ts [hunk 13]
```diff
@@ -111,4 +126,51 @@ export class Seam extends Routes {
...
+      headers["seam-publishable-key"] = ops.publishedKey
+    } else if (ops.apiKey?.startsWith("seam_")) {
+      // backend mode
+      headers["seam-api-key"] = ops.apiKey
+    }
+    headers["seam-user-identifier-key"] = ops.userIdentifierKey
+
+    try {
+      const response = await axios.post(
+        ops.endpoint + "internal/client_sessions/create",
```
Comment (razor-x) [inline src/seam-connect/client.ts:146]: We need to make sure we use the same option paring logic / defaults here. Just tested this locally w/out passing the endpoint and got errors like this

```
r.js:210     POST http://localhost:5173/undefinedinternal/client_sessions/create 404 (Not Found)
dispat
```

File: src/seam-connect/client.ts [hunk 14]
```diff
@@ -52,33 +66,67 @@ export class Seam extends Routes {
...
-        "SEAM_API_KEY not found in environment, and apiKey not provided"
+    const {
+      apiKey,
+      endpoint,
+      workspaceId,
+      axiosOptions,
+      clientAccessToken,
+      clientSessionToken,
+    } = getSeamClientOptionsWithDefaults(apiKeyOrOptions)
+    let bearer = `Bearer `
```
Comment (razor-x) [inline src/seam-connect/client.ts:77]: This large nested if/else block is hard to parse. Perhaps easier to maintain if we move this to a function, then we can do early return. 🐧 E.g.,

```
const getBearerToken({ clientSessionToken, apiKey }) => {
   if (apiKey || clientAccessToken) {
     throw new Error(
        "You can't use clientSessionToken AND specify apiKey or clientAccessToken."
      )
    }
    if (clientSessionToken} {
      if (!clientSessionToken?.startsWith("seam_cst")) throw
      return clientSessionToken
    }
```

File: src/seam-connect/client.ts [hunk 15]
```diff
@@ -111,4 +159,45 @@ export class Seam extends Routes {
...
+      // backend mode
+      headers["seam-api-key"] = ops.apiKey
+    }
+    headers["seam-user-identifier-key"] = ops.userIdentifierKey
+
+    try {
+      const response: AxiosResponse & {
+        data: { client_session: ClientSession }
+      } = await axios.post(
+        (ops.endpoint ?? "https://connect.getseam.com") +
```
Comment (razor-x) [inline src/seam-connect/client.ts:181]: Can we re-use `getSeamClientOptionsWithDefaults` here? We are missing the fallback to `SEAM_API_URL`.

Comment (razor-x) [inline src/seam-connect/client.ts:181]: Can we normalize `endpoint` so we always strip any trailing slashes?

File: src/seam-connect/client.ts (hunk 7 above)
Comment (razor-x) [reply src/seam-connect/client.ts:164]: That's ok to have it's own type the, but we should make sure to accept and handle the options that deal with the "global" client behavior. I think it is only missing `axiosOptions`.

File: src/seam-connect/client.ts [hunk 16]
```diff
@@ -111,4 +159,45 @@ export class Seam extends Routes {
...
+    let headers: AxiosRequestHeaders = {}
+
+    if (ops.publishableKey?.startsWith("seam_pk")) {
+      // frontend mode
+      headers["seam-publishable-key"] = ops.publishableKey
+    } else if (ops.apiKey?.startsWith("seam_")) {
+      // backend mode
+      headers["seam-api-key"] = ops.apiKey
+    }
+    headers["seam-user-identifier-key"] = ops.userIdentifierKey
```
Comment (razor-x) [inline src/seam-connect/client.ts:175]:
```suggestion
    if (ops.userIdentifierKey) {                                                                                                                                                                                                                                                       
      headers["seam-user-identifier-key"] = ops.userIdentifierKey                                                                                                                                                                                                                      
    }
```

File: src/seam-connect/client.ts [hunk 17]
```diff
@@ -1,13 +1,27 @@
...
+  SuccessfulAPIResponse,
+} from "../types/globals"
 import { version } from "../../package.json"
+import { ClientSession, ClientSessionResponse } from "../types"
 
 export interface SeamClientOptions {
   /* Seam API Key */
   apiKey?: string
+  /* Seam Client Access Token */
+  clientAccessToken?: string
```
Comment (seveibar) [inline src/seam-connect/client.ts:22]: `clientAccessToken` is mislabeled, `seam_at` should never be used on a client. Maybe `personalAccessToken`?

File: src/seam-connect/client.ts (hunk 14 above)
Comment (seveibar) [reply src/seam-connect/client.ts:77]: agree that this function is a bit hard to parse and early return would work better (especially because this `bearer` var is being mutated)

**Review by seveibar — APPROVED** [review]
agree w/ @razor-x 's comments, also would rename `clientAccessToken` to make sure nobody accidentally uses it on a client.

File: src/seam-connect/client.ts (hunk 16 above)
Comment (razor-x) [reply src/seam-connect/client.ts:175]: I can't remember exactly but this was either generating a type error or a runtime error upstream when `userIdentifierKey` was null or undefined.

File: src/seam-connect/client.ts (hunk 17 above)
Comment (razor-x) [reply src/seam-connect/client.ts:22]: I think this is just the old name right? Should be `clientSessionToken`?

**Review by razor-x — APPROVED** [review]
Looks good, just noted the one issue with not passing though the `axiosOptions`.

File: src/seam-connect/client.ts [hunk 18]
```diff
@@ -111,4 +151,51 @@ export class Seam extends Routes {
...
+      // backend mode
+      headers["seam-api-key"] = apiKey
+    }
+    if (ops.userIdentifierKey) {
+      headers["seam-user-identifier-key"] = ops.userIdentifierKey
+    } else {
+      throw new Error("userIdentifierKey is required")
+    }
+    try {
+      const response: AxiosResponse & {
```
Comment (razor-x) [inline src/seam-connect/client.ts:177]: @azat-co I didn't save my comment 😭 

Anyway, was just going to say we should make sure to respect `axiosOptions`, so like in the constructor, something like

```ts
    const client = axios.create({
      ...axiosOptions,
      baseURL: endpoint,
      headers,
    })

   await client.post(...)
```


Looking closer, I just realized the normal client behavior does a bit more, like always throw `SeamAPIError`. We should factor out the client creation method and use it both places. I will make an issue for that.

Comment (razor-x) [reply src/seam-connect/client.ts:177]: https://github.com/seamapi/javascript/issues/184

Comment (razor-x) [reply src/seam-connect/client.ts:177]: Yes, basically we should follow the same logic as the constructor for how we use `axios`. I created a [new issue that better explains the desired behavior](https://github.com/seamapi/javascript/issues/184). It's ok if we resolve that issue in a follow-up PR.
